bindings/profilers/wall.cc

```diff
@@ -25,208 +26,413 @@
+
+  LabelSetsByNode labelSetsByNode;
+
+  if (contexts.empty()) {
+    return labelSetsByNode;
+  }
+  auto isolate = Isolate::GetCurrent();
+
+  SampleContext sampleContext = contexts.pop_front();
+
+  uint64_t time_diff = last_start - profile->GetStartTime() * 1000;
+
+  for (int i = 0; i < profile->GetSamplesCount(); i++) {
+    auto sample = profile->GetSample(i);
+    if (isIdleSample(sample)) {
+      continue;
+    }
+
+    // Translate from profiler micros to uv_hrtime nanos
+    auto sampleTimestamp = profile->GetSampleTimestamp(i) * 1000 + time_diff;
+
+    // This loop will drop all contexts that are too old to be associated with
+    // the current sample; associate those (normally one) that are close enough
+    // to it in time, and stop as soon as it sees a context that's too recent
+    // for this sample.
+    for (;;) {
+      auto contextTimestamp = sampleContext.timestamp;
+      if (contextTimestamp < sampleTimestamp - halfInterval) {
+        // Current sample context is too old, discard it and fetch the next one
+        if (contexts.empty()) {
+          return labelSetsByNode;
+        }
+        sampleContext = contexts.pop_front();
+      } else if (contextTimestamp >= sampleTimestamp + halfInterval) {
+        // Current sample context is too recent, we'll try to match it to the
+        // next sample
+        break;
+      } else {
+        // This context timestamp is in the goldilocks zone around the sample
+        // timestamp, so associate its labels with the sample.
+        auto it = labelSetsByNode.find(sample);
+        Local<Array> array;
+        if (it == labelSetsByNode.end()) {
+          array = Nan::New<Array>();
+          labelSetsByNode.emplace(sample, array);
+        } else {
+          array = it->second;
+        }
+        Nan::Set(
+            array, array->Length(), sampleContext.labels.get()->Get(isolate));
+        // Sample context was consumed, fetch the next one
+        if (contexts.empty()) {
+          return labelSetsByNode;
+        }
+        sampleContext = contexts.pop_front();
+      }
+    }
+  }
+  // Push the last popped sample context back into the ring to be used by the
+  // next profile
+  contexts.push_front(std::move(sampleContext));
+  return labelSetsByNode;
 }
 
-WallProfiler::WallProfiler(int interval) : samplingInterval(interval) {}
+static Nan::Persistent<v8::Function> constructor;
+
+WallProfiler::WallProfiler(int intervalMicros, int durationMicros)
+    : samplingInterval(intervalMicros),
+      contexts(durationMicros * 2 / intervalMicros) {}
 
 WallProfiler::~WallProfiler() {
-  Dispose();
+  Dispose(nullptr);
+}
+
+template <typename F>
+void updateProfilers(F updateFn) {
+  auto currProfilers = profilers.load();
+  for (;;) {
+    auto newProfilers = new ProfilerMap(*currProfilers);
+    updateFn(newProfilers);
+    if (profilers.compare_exchange_weak(currProfilers, newProfilers)) {
+      delete currProfilers;
```

Comment:
Made some changes, using atomic sets to `nullptr` for the sighandler to temporarily take ownership of the map while it's using it

Reply:
Not sure about this, it seems that it does not solve the issue that when deleting `currProfilers` on line 403 another thread might be copying it on 399 because it loaded its value before CAS operation succeeded on line 401.